### ISSUE TITLE
Teach Drasil about ODEs with events

### ISSUE THREAD

**Issue by cd155** [JacquesCarette/Drasil#2878]
In the #2875 , we have the NewDEModel containing ODE data represent. One concept to can be added is ODE with events. 

For example, in the SWHS, it has multiple cases on the right-hand side.
<img src="https://user-images.githubusercontent.com/16947266/139307771-4ea96ac9-e7c4-4d9b-a85b-1418deaa6a37.png" width="500">

In the current example, it uses `completeCase` to handle it, `completeCase` is a `relation` type.
```haskell
htPCMRel = sy pcmE $= completeCase [case1, case2, case3]
  where ...
```

**Comment by cd155** [comment]
To record the feedback about ODE with events from Dr.Smith

The [IM:eBalanceOnPCM](https://github.com/JacquesCarette/Drasil/blob/10f9f9ed0e332c7a31875f6e035db96549c45945/code/drasil-example/swhs/lib/Drasil/SWHS/IMods.hs#L155) ODE has three cases because the PCM heats differently depending on whether it is solid, melting, or all liquid.  The cases where it is all solid or all liquid are pretty easy to understand.  The thermal related material properties change, but the form of the ODE is the same.  The case that is different is when the PCM is melting.  In this case dT_P/dt = 0 because all of the thermal energy that is added is going to melting the PCM.  No energy is “left” for raising the temperature, so the temperature at the beginning of melting (the melting point) and at the end of melting is the same.  Once all the PCM is melted, the added energy will go toward raising the temperature of the liquid PCM.

Cases like this are considered to be ODEs with events.  When the event occurs the form of the ODE changes.  A good overview of this is given here:

https://www.sciencedirect.com/science/article/pii/S0898122100000456

If we want to “teach” Drasil about ODEs with events, then we need to add the notion of an associated event function.  In the paper I pointed to above, the event function is g(t, y).  An event occurs at t* when g(t*, y(t*)) = 0.  

The condition for the first “switch” is when T_P < T_melt.  The function g in this case would be g = T_melt - T_P.  (You can see this in the Matlab code for SWHS here: https://github.com/smiths/swhs/blob/master/src/Program/Matlab/event1.m).

It seems to me that teaching Drasil about events for ODEs wouldn’t be trivial.  We have to introduce the function g as a parameter.  Moreover, we need to consider cases where there is more than one event function (like SWHS).  The direction that the condition occurs from is also relevant.  We only heat the PCM, but in general, the ODE would be different if it is heating or cooling.  You’ll also see that the “canonical” form of g is different from how people usually communicate equations.  This means the “model" and the display of the model will be different.

We should definitely teach Drasil about ODEs with events, but I think for now, we should follow Jacques' suggestion of continuing to use “other” as the model type for SWHS.  Drasil just doesn’t know enough yet to understand an ODE with events.

**Comment by JacquesCarette** [comment]
I agree, teaching Drasil about ODEs with events is beyond the scope of your work. Personally, I would actually change SWHS back to `OthModel`. This will let you delete `DEModel`.  @smiths ?

**Comment by smiths** [comment]
Yes, I agree.

**Comment by balacij** [comment]
Out of curiosity, what is meant by "events"? I'm not quite sure I fully understand.

**Comment by smiths** [comment]
An event is when the ODE changes.  A classic example is a bouncing ball.  You drop the ball from a given height.  The event occurs when the ball hits the ground.  Once the ball hits the ground the initial conditions change because when the ball bounces up the velocity in the y-direction has changed direction.  Another example is modelling the temperature in a room with a thermostat connected to a heat source.  When the temperature gets too low, the heat source comes and the ODE changes.  SWHS is another example.  There is one ODE when the PCM is solid, another when it is melting, and another when it is fully melted.  When there are events the ODE solver needs to determine when the event occurs.  For instance, the solver needs to determine when the ball hits the ground.

A mathematical characterization of event location for ODEs can be found in [Shampine and Thompson 1999](https://gitlab.cas.mcmaster.ca/smiths/pub/-/blob/master/ShampineAndThompson2000.pdf).  [Matlab](https://www.mathworks.com/help/matlab/math/ode-event-location.html) supports ODEs with events.
